// @ali/tingle-textarea-field/src/TextareaField.js
/**
 * TextareaField Component for tingle
 * @author zhangshun@alipay.com
 *
 * Copyright 2014-2016, Tingle Team.
 * All rights reserved.
 */
const React = require('react');
const classnames = require('classnames');
const Context = require('@ali/tingle-context');
const calculateHeight = require('./calculateHeight');
const Field = require('@ali/tingle-field');

let prefixClass = Context.prefixClass;

class TextareaField extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            height: null,
            rows:1
        };
    }

    componentDidMount() {
        this._resize();
    }

    componentWillReceiveProps(nextProps) {
        if ('value' in nextProps) {
            this._resize();
        }
    }

    _resize() {
        this.setState(
            calculateHeight(ReactDOM.findDOMNode(this.refs.textarea), this.props.minRows || this.props.rows, this.props.maxRows)
        );
    }

    handleChange(e) {
        this._resize();
        this.props.onChange(e.target.value, e);
    }

    handleFocus(e) {
        this.props.onFocus(e);
    }

    handleBlur(e) {
        this.props.onBlur(e);
    }

    render() {
        let t = this;
        let { placeholder, label, readOnly, lineHeight } = t.props;
        let style = {
            //height: t.state.height,
            lineHeight:lineHeight
        };

        return (
            <Field {...t.props} multiLine={true} className={classnames({
                [prefixClass('textarea-field')]: true,
                readonly: readOnly,
                [t.props.className]: !!t.props.className
            })}>
                <textarea ref="textarea"
                    className={prefixClass('textarea-field-content')}
                    style={style}
                    placeholder={placeholder}
                    value={t.props.value}
                    readOnly={readOnly}
                    rows={t.state.rows}
                    onChange={t.handleChange.bind(t)}
                    onFocus={t.handleFocus.bind(t)}
                    onBlur={t.handleBlur.bind(t)}/>
            </Field>
        );
    }
};

TextareaField.defaultProps = {
    placeholder: '',
    onChange: Context.noop,
    onFocus: Context.noop,
    onBlur: Context.noop,
    readOnly: false,
    minRows: 1,
    maxRows: 10,
    lineHeight:'1.3',
    value: ''
};

// http://facebook.github.io/react/docs/reusable-components.html
TextareaField.propTypes = {
    value: React.PropTypes.string,
    placeholder: React.PropTypes.string,
    onChange: React.PropTypes.func,
    onFocus: React.PropTypes.func,
    onBlur: React.PropTypes.func,
    readOnly: React.PropTypes.bool,
    minRows: React.PropTypes.number,
    maxRows: React.PropTypes.number,
    lineHeight: React.PropTypes.string,
    className: React.PropTypes.string
};

TextareaField.displayName = 'TextareaField';

module.exports = TextareaField;
