// @ali/tingle-mask/dist/Mask.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Mask Component for tingle
 * @author quanyun.mqy
 *
 * Copyright 2014-2016, Tingle Team.
 * All rights reserved.
 */
var React = require('react');
var classnames = require('classnames');

var _require = require('@ali/tingle-context');

var prefixClass = _require.prefixClass;
var noop = _require.noop;

var Mask = function (_React$Component) {
    _inherits(Mask, _React$Component);

    function Mask(props) {
        _classCallCheck(this, Mask);

        var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(Mask).call(this, props));

        _this.state = {
            visible: props.visible
        };
        return _this;
    }

    _createClass(Mask, [{
        key: 'handleClick',
        value: function handleClick() {
            var t = this;
            if (t.props.closeable === false || t.props.onWillHide() === false) {
                return;
            }
            t.setState({
                visible: false
            }, function () {
                t.props.onDidHide();
            });
        }
    }, {
        key: 'componentWillReceiveProps',
        value: function componentWillReceiveProps(nextProps) {
            this.setState({
                visible: nextProps.visible
            });
        }
    }, {
        key: 'render',
        value: function render() {
            var t = this;
            var _t$props = t.props;
            var className = _t$props.className;
            var opacity = _t$props.opacity;
            var zIndex = _t$props.zIndex;
            var style = _t$props.style;

            var other = _objectWithoutProperties(_t$props, ['className', 'opacity', 'zIndex', 'style']);

            var visible = t.state.visible;


            return React.createElement('div', _extends({ ref: 'root', className: classnames(prefixClass('mask'), _defineProperty({
                    visible: visible
                }, className, !!className)), style: {
                    backgroundColor: 'rgba(19, 21, 26, ' + opacity + ')',
                    opacity: visible ? 1 : 0,
                    zIndex: zIndex
                }, onClick: t.handleClick.bind(t) }, other));
        }
    }]);

    return Mask;
}(React.Component);

Mask.defaultProps = {
    closeable: true,
    opacity: 0.6,
    onDidHide: noop,
    onWillHide: noop,
    visible: false,
    zIndex: 900
};

// http://facebook.github.io/react/docs/reusable-components.html
Mask.propTypes = {
    className: React.PropTypes.string,
    closeable: React.PropTypes.bool,
    opacity: React.PropTypes.number,
    onDidHide: React.PropTypes.func,
    onWillHide: React.PropTypes.func,
    visible: React.PropTypes.bool,
    zIndex: React.PropTypes.number
};

Mask.displayName = 'Mask';

module.exports = Mask;
