// @ali/tingle-field/dist/Field.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Field Component for tingle
 * @author gnosaij
 *
 * Copyright 2014-2016, Tingle Team.
 * All rights reserved.
 */
var React = require('react');
var classnames = require('classnames');
var Icon = require('@ali/tingle-icon');
var Context = require('@ali/tingle-context');

var prefixClass = Context.prefixClass;

var Field = function (_React$Component) {
    _inherits(Field, _React$Component);

    function Field(props) {
        _classCallCheck(this, Field);

        return _possibleConstructorReturn(this, Object.getPrototypeOf(Field).call(this, props));
    }

    _createClass(Field, [{
        key: 'render',
        value: function render() {
            var _classnames2;

            var t = this;
            var requiredTag = React.createElement(Icon, { name: 'field-required', className: prefixClass('field-layout-label-required'), width: 6, height: 6, fill: 'red' });
            return React.createElement(
                'div',
                { className: classnames(prefixClass('field'), _defineProperty({}, t.props.className, !!t.props.className)) },
                t.props.label && t.props.layout === 'v' && React.createElement(
                    'div',
                    { className: prefixClass('field-layout-v-label') },
                    t.props.label,
                    this.props.required && requiredTag
                ),
                React.createElement(
                    'div',
                    { className: classnames(prefixClass('field-box FBH'), (_classnames2 = {}, _defineProperty(_classnames2, prefixClass('TE'), t.props.tappable), _defineProperty(_classnames2, prefixClass('FBAC'), !t.props.multiLine), _classnames2)) },
                    t.props.label && t.props.layout === 'h' && React.createElement(
                        'div',
                        { className: prefixClass('field-layout-h-label') },
                        t.props.label,
                        this.props.required && requiredTag
                    ),
                    React.createElement(
                        'div',
                        { className: classnames(prefixClass('FB1 PR'), _defineProperty({}, prefixClass('field-multi'), t.props.multiLine)) },
                        t.props.children
                    ),
                    t.props.icon && t.props.icon.name && React.createElement(
                        'div',
                        { className: prefixClass('FBH FBAC field-icon') },
                        React.createElement(Icon, t.props.icon)
                    )
                )
            );
        }
    }]);

    return Field;
}(React.Component);

Field.defaultProps = {
    label: '',
    tappable: false,
    required: false,
    multiLine: false,
    icon: {},
    layout: 'h'
};

// http://facebook.github.io/react/docs/reusable-components.html
Field.propTypes = {
    label: React.PropTypes.string,
    icon: React.PropTypes.object,
    required: React.PropTypes.bool,
    tappable: React.PropTypes.bool,
    multiLine: React.PropTypes.bool,
    layout: React.PropTypes.oneOf(['h', 'v'])
};

Field.displayName = 'Field';

module.exports = Field;
