// @ali/tingle-tab-bar/dist/TabBarItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _tingleContext = require('@ali/tingle-context');

var _tingleContext2 = _interopRequireDefault(_tingleContext);

var _tingleBox = require('@ali/tingle-box');

var _tingleIcon = require('@ali/tingle-icon');

var _tingleIcon2 = _interopRequireDefault(_tingleIcon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * TabBar Component for tingle
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author zhouwenjie
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2014-2016, Tingle Team.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


var isUrl = function isUrl(str) {
    return (/\/\/.+/g.test(str)
    );
};
var isArray = function isArray(obj) {
    return Object.prototype.toString.call(obj) === '[object Array]';
};

var TabBarItem = function (_React$Component) {
    _inherits(TabBarItem, _React$Component);

    function TabBarItem(props) {
        _classCallCheck(this, TabBarItem);

        return _possibleConstructorReturn(this, Object.getPrototypeOf(TabBarItem).call(this, props));
    }

    _createClass(TabBarItem, [{
        key: 'renderImgIcon',
        value: function renderImgIcon(src, title) {
            var className = _tingleContext2.default.prefixClass('tab-bar-item-img-icon');
            return React.createElement('img', { className: className, src: src, alt: title });
        }
    }, {
        key: 'renderTingleIcon',
        value: function renderTingleIcon(name) {
            var className = _tingleContext2.default.prefixClass('tab-bar-item-icon');
            return React.createElement(_tingleIcon2.default, { width: 25, height: 25, className: className, name: name });
        }
    }, {
        key: 'renderIcon',
        value: function renderIcon(icon, title, active) {
            if (typeof icon === 'string') {
                if (isUrl(icon)) {
                    return this.renderImgIcon(icon, title);
                }
                return this.renderTingleIcon(icon);
            }

            if (isArray(icon)) {
                return this.renderImgIcon(active ? icon[1] : icon[0], title);
            }
            return null;
        }
    }, {
        key: 'render',
        value: function render() {
            var t = this;
            var item = t.props.item;
            var _item$props = item.props;
            var title = _item$props.title;
            var icon = _item$props.icon;

            var active = t.props.active;
            return React.createElement(
                _tingleBox.Box,
                { className: (0, _classnames2.default)(_tingleContext2.default.prefixClass('tab-bar-item'), {
                        'active': active
                    }), onClick: t.props.onClick, flex: 1
                },
                t.renderIcon(icon, title, active),
                React.createElement(
                    'span',
                    { className: _tingleContext2.default.prefixClass('tab-bar-item-label') },
                    title
                )
            );
        }
    }]);

    return TabBarItem;
}(React.Component);

TabBarItem.displayName = 'TabBarItem';
TabBarItem.defaultProps = {};
TabBarItem.propTypes = {
    className: React.PropTypes.string,
    item: React.PropTypes.element
};
exports.default = TabBarItem;
