// @ali/tingle-button/dist/Button.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
* Button Component for tingle
* @author cm
*
* Copyright 2014-2016, Tingle Team, Alinw.
* All rights reserved.
*/
var React = require('react');
var classnames = require('classnames');
var Context = require('@ali/tingle-context');

var prefixClass = function prefixClass(name) {
    return Context.prefixClass ? Context.prefixClass(name) : 't-' + name;
};

var Button = function (_React$Component) {
    _inherits(Button, _React$Component);

    function Button(props) {
        _classCallCheck(this, Button);

        return _possibleConstructorReturn(this, (Button.__proto__ || Object.getPrototypeOf(Button)).call(this, props));
    }

    _createClass(Button, [{
        key: 'handleClick',
        value: function handleClick(evt) {
            if (!this.props.disabled) {
                this.props.onClick(evt);
            }
        }
    }, {
        key: 'render',
        value: function render() {
            var _classSet;

            var size = this.props.size;

            var type = this.props.type;
            var primary = type === 'primary';
            var secondary = type === 'secondary';
            var danger = type === 'danger';
            var text = type === 'text';
            var minor = type === 'minor';

            var disabled = !!this.props.disabled;
            var classSet = (_classSet = {}, _defineProperty(_classSet, '' + prefixClass('FBH FBAC FBJC'), true), _defineProperty(_classSet, this.props.className, !!this.props.className), _defineProperty(_classSet, '' + prefixClass('button'), true), _defineProperty(_classSet, 'disabled', disabled), _defineProperty(_classSet, '' + prefixClass('button-small'), size === 'small'), _defineProperty(_classSet, '' + prefixClass('button-large'), size === 'large'), _defineProperty(_classSet, '' + prefixClass('button-primary'), primary && !disabled), _defineProperty(_classSet, '' + prefixClass('button-secondary'), secondary && !disabled), _defineProperty(_classSet, '' + prefixClass('button-minor'), minor && !disabled), _defineProperty(_classSet, '' + prefixClass('button-danger'), danger && !disabled), _defineProperty(_classSet, '' + prefixClass('button-text'), text), _classSet);

            var style = this.props.style;

            return React.createElement(
                'div',
                { className: classnames(classSet),
                    disabled: disabled,
                    style: style,
                    onClick: this.handleClick.bind(this) },
                this.props.children
            );
        }
    }]);

    return Button;
}(React.Component);

Button.defaultProps = {
    size: 'medium',
    type: 'primary',
    disabled: false,
    style: {},
    onClick: function onClick() {}
};

// http://facebook.github.io/react/docs/reusable-components.html
Button.propTypes = {
    size: React.PropTypes.oneOf(['medium', 'large', 'small']),
    type: React.PropTypes.oneOf(['primary', 'secondary', 'minor', 'danger', 'text']),
    disabled: React.PropTypes.bool,
    style: React.PropTypes.object,
    onClick: React.PropTypes.func
};

Button.displayName = 'Button';

module.exports = Button;
