// @ali/tingle-button/dist/prototype.js
'use strict';

var React = require('react');

var _require = require('engine');

var Bundle = _require.Bundle;

var _require2 = require('engine-utils');

var BoolSetter = _require2.BoolSetter;
var TextSetter = _require2.TextSetter;
var ChoiceSetter = _require2.ChoiceSetter;
var JsonSetter = _require2.JsonSetter;
var NumberSetter = _require2.NumberSetter;


module.exports = Bundle.createPrototype({
    title: " 按钮",
    category: "*",
    icon: require("./logo-ide.svg"),
    componentName: "Button",
    canHovering: true,
    canSelecting: true,
    canDraging: true,
    isInline: false,
    isContainer: false,
    canDropto: true,
    canDroping: false,
    configure: [{
        name: "size",
        title: "size",
        defaultValue: 'medium',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(ChoiceSetter, { options: [{ value: 'medium' }, { value: 'large' }, { value: 'small' }] })
    }, {
        name: "type",
        title: "type",
        defaultValue: 'primary',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(ChoiceSetter, { options: [{ value: 'primary' }, { value: 'secondary' }, { value: 'minor' }, { value: 'danger' }, { value: 'text' }] })
    }, {
        name: "disabled",
        title: "disabled",
        defaultValue: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(BoolSetter, null)
    }, {
        name: "style",
        title: "style",
        defaultValue: {},
        required: false,
        fieldStyle: "accordion",
        fieldCollapsed: false,
        setter: React.createElement(JsonSetter, null)
    }, {
        name: "children",
        title: "text",
        defaultValue: "一级按钮",
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(TextSetter, null)
    }]
});
