// @ali/tingle-text-field/dist/prototype.js
'use strict';

var React = require('react');

var _require = require('engine'),
    Bundle = _require.Bundle;

var _require2 = require('engine-utils'),
    BoolSetter = _require2.BoolSetter,
    TextSetter = _require2.TextSetter,
    ChoiceSetter = _require2.ChoiceSetter,
    JsonSetter = _require2.JsonSetter,
    NumberSetter = _require2.NumberSetter;

module.exports = Bundle.createPrototype({
    title: "单行文本",
    category: "*",
    icon: require("./logo-ide.svg"),
    componentName: "TextField",
    canHovering: true,
    canSelecting: true,
    canDraging: true,
    isInline: true,
    isContainer: false,
    canDropto: true,
    canDroping: false,
    configure: [{
        name: "label",
        title: "label",
        defaultValue: '姓名',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(TextSetter, { multiline: true, rows: 2 })
    }, {
        name: "placeholder",
        title: "placeholder",
        defaultValue: '',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(TextSetter, { multiline: true, rows: 2 })
    }, {
        name: "readOnly",
        title: "readOnly",
        defaultValue: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(BoolSetter, null)
    }, {
        name: "type",
        title: "type",
        defaultValue: 'text',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(TextSetter, { multiline: true, rows: 2 })
    }, {
        name: "value",
        title: "value",
        defaultValue: '小明',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(TextSetter, { multiline: true, rows: 2 })
    }, {
        name: "className",
        title: "className",
        defaultValue: '',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(TextSetter, { multiline: true, rows: 2 })
    }]
});
