// @ali/tingle-text-field/dist/TextField.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * TextField Component for tingle
 * @author
 *
 * Copyright 2014-2016, Tingle Team.
 * All rights reserved.
 */
var React = require('react');
var classnames = require('classnames');
var Context = require('@ali/tingle-context');
var Field = require('@ali/tingle-field');

var prefixClass = Context.prefixClass;

var TextField = function (_React$Component) {
    _inherits(TextField, _React$Component);

    function TextField(props) {
        _classCallCheck(this, TextField);

        return _possibleConstructorReturn(this, (TextField.__proto__ || Object.getPrototypeOf(TextField)).call(this, props));
    }

    _createClass(TextField, [{
        key: 'handleChange',
        value: function handleChange(e) {
            var value = this.props.filter(e.target.value);
            this.props.onChange(value, e);
        }
    }, {
        key: 'handleFocus',
        value: function handleFocus(e) {
            var t = this;
            t.props.onFocus(t.props.value);
        }
    }, {
        key: 'handleBlur',
        value: function handleBlur(e) {
            var t = this;
            t.props.onBlur(t.props.value, e);
        }
    }, {
        key: 'render',
        value: function render() {
            var t = this;

            return React.createElement(
                Field,
                _extends({}, t.props, { className: classnames(prefixClass('text-field'), t.props.className, {
                        readonly: t.props.readOnly
                    }) }),
                React.createElement(
                    'div',
                    { ref: 'placeholder', className: classnames(prefixClass('omit text-field-placeholder'), _defineProperty({}, prefixClass('DN'), t.props.value !== '')) },
                    t.props.placeholder
                ),
                React.createElement('input', { ref: 'input', className: prefixClass('text-field-input'),
                    type: t.props.type, value: t.props.value, readOnly: t.props.readOnly,
                    onChange: t.handleChange.bind(t),
                    onFocus: t.handleFocus.bind(t),
                    onBlur: t.handleBlur.bind(t) })
            );
        }
    }]);

    return TextField;
}(React.Component);

TextField.defaultProps = {
    className: '',
    filter: function filter(v) {
        return v;
    },
    label: '',
    onChange: Context.noop,
    onFocus: Context.noop,
    onBlur: Context.noop,
    placeholder: '',
    readOnly: false,
    type: 'text',
    value: ''
};

TextField.propTypes = {
    className: React.PropTypes.string,
    filter: React.PropTypes.func,
    label: React.PropTypes.string,
    onChange: React.PropTypes.func,
    onFocus: React.PropTypes.func,
    onBlur: React.PropTypes.func,
    placeholder: React.PropTypes.string,
    readOnly: React.PropTypes.bool,
    type: React.PropTypes.string,
    value: React.PropTypes.string
};

TextField.displayName = 'TextField';

module.exports = TextField;
