// @ali/tingle-checkbox-field/src/CheckboxField.js
/**
 * CheckboxField Component for tingle
 * @author shanchao
 *
 * Copyright 2014-2016, Tingle Team.
 * All rights reserved.
 */
const React = require('react');
const classnames = require('classnames');
const Icon = require('@ali/tingle-icon');
const Context = require('@ali/tingle-context');
const Group = require('@ali/tingle-group');
const Field = require('@ali/tingle-field');
const SelectLayer = require('./SelectLayer');
const prefixClass = Context.prefixClass;

class CheckboxField extends React.Component {

    constructor(props) {
        super(props);

        const t = this;
        t.state = {
            selectedText: ''
        };
    }

    componentWillReceiveProps(nextProps) {
        this.getSelectedText(nextProps.data);
    }

    getSelectedText(data) {
        let selectedText = '';
        data.forEach( item => {
            if (item.checked) {
                if(this.props.mode === 'list') {
                    selectedText = selectedText + '; ' + item.text;
                } else {
                    selectedText = selectedText + '; ' + (item.slotText ? item.slotText : item.text);
                }
            }
        });

        if (selectedText) {
            selectedText = selectedText.substring(2);
        }

        this.state.selectedText = selectedText;
    }

    componentWillMount() {
        this.getSelectedText(this.props.data);
    }

    getData() {
        let t = this;
        let data = [];

        t.props.data.forEach( function (item) {
            if (item.checked) {
                data.push(item);
            }
        });

        return data;
    }

    clickAction(value, item, index, data) {
        let t = this;
        let {onChange} = t.props;
        let disable = item.disable;
        if (t.props.readOnly || disable) {
            return;
        }
        item.checked = !item.checked;
        onChange && onChange(t.getData());
        t.forceUpdate();
    }

    renderList() {
        let t = this;
        let props = t.props;
        let {className, data:checkboxArray, groupListArgument, groupListFlag, label} = props;
        const requiredTag = (
            <Icon
                name="field-required"
                className={prefixClass('field-layout-label-required')}
                width={26}
                height={26}
                fill="red"
            />
        );

        let checkboxArrayComponent = checkboxArray.map(function (item, index, data) {
            let {checked,disable, value} = item;
            let iconName = checked ? "check-round" : "";
            let iconClassName = classnames(
                prefixClass("checkbox-field-icon"),
                {"checked": checked},
                {"noChecked": !checked},
                {"disable": disable},
                {[prefixClass("checkbox-field-icon") + '-list']: true}
            );

            let textClassName = classnames(
                prefixClass("checkbox-field-text FBH FBAC"),
                {"disable": disable}
            );

            let iconTag = checked ?
                <Icon
                    key="check-round"
                    width={26}
                    height={26}
                    name="check-round"
                    className={iconClassName}
                /> : <div className={iconClassName}></div>;

            let finalItemJSX = (
                <div
                    onClick={t.clickAction.bind(t, value, item,index, checkboxArray)}
                    key={index}
                    className={prefixClass("checkbox-field-row FBAC FBH")}
                >
                    <div className={prefixClass("checkbox-field-icon-div FBH FBAC")}>
                        {iconTag}
                    </div>
                    <div ref={"text" + index} className={textClassName}>{item.text}</div>
                </div>

            );
            return (
                finalItemJSX
            );
        });


        let finalJSX = (
            <Group className={classnames(prefixClass('checkbox-field'), {
                [className]: !!className,
                [prefixClass('checkbox-field-readonly')]: !!t.props.readOnly
            })}>
                {
                    label == '' ? null :
                        <Group.Head className={classnames(prefixClass('checkbox-label'))}>
                            {label}
                            {this.props.required && requiredTag}
                        </Group.Head>
                }
                <Group.List {...groupListArgument} >
                    {checkboxArrayComponent}
                </Group.List>
            </Group>
        );

        if (!groupListFlag) {
            finalJSX = (
                <div ref='root' className={classnames(prefixClass('checkbox-field'), {
                    [className]: !!className,
                    [prefixClass('checkbox-field-readonly')]: !!t.props.readOnly
                })}>
                    {checkboxArrayComponent}
                </div>
            )
        }
        return finalJSX;
    }

    handleClick() {
        !this.props.readOnly && this.refs.slot.show();
    }

    handleCancel() {
    }

    handleConfirm(data) {
        this.state.value = data;
        this.getSelectedText(data);
        this.setState(this.state);
        this.props.onChange(data);
    }

    renderSlot() {
        const t = this;
        return (
            <Field
                {...t.props}
                icon={{
                    className: classnames(
                        prefixClass('checkbox-field-icon'),
                        {[prefixClass("checkbox-field-icon") + '-slot']: true}
                    ),
                    name: 'angle-right',
                    width: 26,
                    height: 26,
                    onClick: t.handleClick.bind(t)
                }}
                className={classnames(prefixClass('checkbox-field'), {
                    [t.props.className]: !!t.props.className
                })}>
                <div onClick={t.handleClick.bind(t)} className={prefixClass('checkbox-field-value-wrap')}>
                    {
                        t.state.selectedText ?
                            <div className={prefixClass('checkbox-field-value-list')}>
                                {t.state.selectedText}
                            </div> :
                            <div className={prefixClass('omit checkbox-field-placeholder')}>
                                {t.props.placeholder}
                            </div>
                    }
                </div>
                <SelectLayer ref="slot" title={t.props.label}
                    confirmText={t.props.confirmText}
                    cancelText={t.props.cancelText}
                    data={t.props.data}
                    value={t.state.value}
                    maskCloseable={t.props.maskCloseable}
                    onCancel={t.handleCancel.bind(t)}
                    onConfirm={t.handleConfirm.bind(t)}/>
            </Field>
        )
    }

    render() {
        return this.props.mode === 'list' ? this.renderList() : this.renderSlot();
    }
}

CheckboxField.defaultProps = {
    mode: 'slot', // slot, list
    readOnly: false,
    label:'',
    data: [],
    onChange: function () {
    },
    placeholder: '',
    maskCloseable: true,
    groupListFlag: true,
    groupListArgument: {
        lineIndent: 0,
        itemIndent: 15
    }
};

// http://facebook.github.io/react/docs/reusable-components.html
CheckboxField.propTypes = {
    className: React.PropTypes.string,
    mode: React.PropTypes.string,
    readOnly: React.PropTypes.bool,
    label: React.PropTypes.string,
    data: React.PropTypes.array,
    onChange: React.PropTypes.func,
    placeholder: React.PropTypes.string,
    maskCloseable: React.PropTypes.bool,
    groupListFlag: React.PropTypes.bool,
    groupListArgument: React.PropTypes.object
};

CheckboxField.displayName = 'CheckboxField';

module.exports = CheckboxField;
