// @ali/tingle-text-field/src/prototype.js
const React = require('react');
const {Bundle} = require('engine');
const {BoolSetter, TextSetter, ChoiceSetter, JsonSetter, NumberSetter} = require('engine-utils');

module.exports = Bundle.createPrototype({
    title: "单行文本",
    category: "*",
    icon: require("./logo-ide.svg"),
    componentName: "TextField",
    canHovering: true,
    canSelecting: true,
    canDraging: true,
    isInline: true,
    isContainer: false,
    canDropto: true,
    canDroping: false,
    configure: [{
        name: "label",
        title: "label",
        defaultValue: '姓名',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: <TextSetter multiline={true} rows={2}/>
    }, {
        name: "placeholder",
        title: "placeholder",
        defaultValue: '',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: <TextSetter multiline={true} rows={2}/>
    }, {
        name: "readOnly",
        title: "readOnly",
        defaultValue: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: <BoolSetter />
    }, {
        name: "type",
        title: "type",
        defaultValue: 'text',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: <TextSetter multiline={true} rows={2}/>
    }, {
        name: "value",
        title: "value",
        defaultValue: '小明',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: <TextSetter multiline={true} rows={2}/>
    }, {
        name: "className",
        title: "className",
        defaultValue: '',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: <TextSetter multiline={true} rows={2}/>
    }]
});
