// @ali/tingle-rate/dist/Rate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _tingleStyle = require('@ali/tingle-style');

var _tingleContext = require('@ali/tingle-context');

var _tingleContext2 = _interopRequireDefault(_tingleContext);

var _tingleIcon = require('@ali/tingle-icon');

var _tingleIcon2 = _interopRequireDefault(_tingleIcon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Rate Component for tingle
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author quanyun.mqy
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2014-2016, Tingle Team.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


var Rate = function (_React$Component) {
    _inherits(Rate, _React$Component);

    function Rate() {
        _classCallCheck(this, Rate);

        return _possibleConstructorReturn(this, (Rate.__proto__ || Object.getPrototypeOf(Rate)).apply(this, arguments));
    }

    _createClass(Rate, [{
        key: 'handleItemClick',
        value: function handleItemClick(v) {
            this.props.onChange(v);
        }

        // http://facebook.github.io/react/docs/reusable-components.html

    }, {
        key: 'render',
        value: function render() {
            var t = this;
            var items = [];
            for (var i = 1; i <= t.props.totalScore; i++) {
                var item = React.createElement(
                    'div',
                    { className: (0, _classnames3.default)(_tingleContext2.default.prefixClass('rate-item'), {
                            't-DIB': true,
                            't-PR': true,
                            active: i <= t.props.score
                        }), key: i, onClick: t.handleItemClick.bind(t, i), style: {
                            width: (0, _tingleStyle.unitize)(t.props.height),
                            height: (0, _tingleStyle.unitize)(t.props.height),
                            paddingLeft: (0, _tingleStyle.unitize)(t.props.gap / 2),
                            paddingRight: (0, _tingleStyle.unitize)(t.props.gap / 2)
                        }
                    },
                    React.createElement(_tingleIcon2.default, { name: 'star', fill: '#dbdbdb', className: (0, _classnames3.default)(_tingleContext2.default.prefixClass('rate-icon')) })
                );
                items.push(item);
            }
            return React.createElement(
                'div',
                { ref: 'root', className: (0, _classnames3.default)(_tingleContext2.default.prefixClass('rate'), _defineProperty({
                        't-FBH': true
                    }, t.props.className, !!t.props.className)) },
                items
            );
        }
    }]);

    return Rate;
}(React.Component);

Rate.displayName = 'Rate';
Rate.defaultProps = {
    totalScore: 5,
    score: 0,
    height: 30,
    gap: 10,
    onChange: function onChange() {}
};
Rate.propTypes = {
    className: React.PropTypes.string,
    totalScore: React.PropTypes.number,
    score: React.PropTypes.number,
    height: React.PropTypes.oneOfType([React.PropTypes.number, React.PropTypes.string]),
    gap: React.PropTypes.oneOfType([React.PropTypes.number, React.PropTypes.string]),
    onChange: React.PropTypes.func
};
exports.default = Rate;
module.exports = exports['default'];
