// @ali/tingle-switch/dist/Switch.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Switch Component for tingle
 * @author ruiyang.dry
 *
 * Copyright 2014-2016, Tingle Team, Alinw.
 * All rights reserved.
 */
var classnames = require('classnames');

var Context = require('@ali/tingle-context');

var Switch = function (_React$Component) {
    _inherits(Switch, _React$Component);

    function Switch(props) {
        _classCallCheck(this, Switch);

        return _possibleConstructorReturn(this, Object.getPrototypeOf(Switch).call(this, props));
    }

    _createClass(Switch, [{
        key: 'handleChange',
        value: function handleChange(event) {
            var t = this;
            if (t.props.readOnly) {
                return;
            }
            t.props.onChange(!t.props.on, event);
        }
    }, {
        key: 'render',
        value: function render() {
            var t = this;
            var classSet = _defineProperty({
                "active": t.props.on,
                "readOnly": t.props.readOnly
            }, t.props.className, !!t.props.className);
            return React.createElement(
                'div',
                {
                    className: classnames(Context.prefixClass('switch'), classSet),
                    readOnly: t.props.readOnly,
                    onClick: t.handleChange.bind(this) },
                React.createElement(
                    'div',
                    { className: Context.prefixClass('switch-back') },
                    React.createElement('div', { className: Context.prefixClass('switch-radius') })
                )
            );
        }
    }]);

    return Switch;
}(React.Component);

Switch.defaultProps = {
    on: true,
    onChange: function onChange() {},

    readOnly: false
};

// http://facebook.github.io/react/docs/reusable-components.html
Switch.propTypes = {
    on: React.PropTypes.bool,
    onChange: React.PropTypes.func,
    readOnly: React.PropTypes.bool
};

Switch.displayName = "Switch";

module.exports = Switch;
