// @ali/tingle-radio-field/dist/RadioField.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames5 = require('classnames');

var _classnames6 = _interopRequireDefault(_classnames5);

var _tingleIcon = require('@ali/tingle-icon');

var _tingleIcon2 = _interopRequireDefault(_tingleIcon);

var _tingleContext = require('@ali/tingle-context');

var _tingleContext2 = _interopRequireDefault(_tingleContext);

var _tingleGroup = require('@ali/tingle-group');

var _tingleGroup2 = _interopRequireDefault(_tingleGroup);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * RadioField Component for tingle
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author shanchao
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2014-2016, Tingle Team.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


var prefixClass = _tingleContext2.default.prefixClass;

var RadioField = function (_React$Component) {
  _inherits(RadioField, _React$Component);

  function RadioField(props) {
    _classCallCheck(this, RadioField);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(RadioField).call(this, props));

    _this.state = {};
    return _this;
  }

  _createClass(RadioField, [{
    key: 'clickAction',
    value: function clickAction(value, item, index, data) {
      var t = this;
      var _t$props = t.props;
      var radioArray = _t$props.data;
      var onChange = _t$props.onChange;


      var disable = item.disable;
      if (disable) {
        return;
      }
      radioArray.map(function (radioItem) {
        radioItem.checked = false;
        return radioItem;
      });
      item.checked = !item.checked;
      if (onChange) {
        onChange(value, index, data);
      }
      t.forceUpdate();
    }
  }, {
    key: 'render',
    value: function render() {
      var t = this;
      var _t$props2 = t.props;
      var rootClassName = _t$props2.rootClassName;
      var className = _t$props2.className;
      var radioArray = _t$props2.data;
      var groupListArgument = _t$props2.groupListArgument;
      var groupListFlag = _t$props2.groupListFlag;
      var label = _t$props2.label;


      var radioArrayComponent = radioArray.map(function (item, index, data) {
        var checked = item.checked;
        var disable = item.disable;
        var value = item.value;

        var textClassName = (0, _classnames6.default)(prefixClass('radio-field-text FBAC FBH FB1'), { disable: disable });
        var iconClassName = (0, _classnames6.default)(prefixClass('radio-field-icon'), { noChecked: !checked }, { checked: checked }, { disable: disable });
        return _react2.default.createElement(
          'div',
          { onClick: t.clickAction.bind(t, value, item, index, data),
            key: index, className: prefixClass('radio-field-row FBAC FBH')
          },
          _react2.default.createElement(
            'div',
            { ref: 'text' + index, className: textClassName },
            item.text
          ),
          _react2.default.createElement(
            'div',
            { className: prefixClass('radio-field-icon-div FBAC FBH') },
            _react2.default.createElement(_tingleIcon2.default, { width: 20, height: 20, name: "check", className: iconClassName })
          )
        );
      });

      var finalJSX = _react2.default.createElement(
        _tingleGroup2.default,
        { className: (0, _classnames6.default)(prefixClass('checkbox-field'), _defineProperty({}, rootClassName, !!rootClassName), _defineProperty({}, className, !!className)) },
        label === '' ? null : _react2.default.createElement(
          _tingleGroup2.default.Head,
          { className: (0, _classnames6.default)(prefixClass('checkbox-label')) },
          label
        ),
        _react2.default.createElement(
          _tingleGroup2.default.List,
          groupListArgument,
          radioArrayComponent
        )
      );

      if (!groupListFlag) {
        finalJSX = _react2.default.createElement(
          'div',
          { ref: 'root', className: (0, _classnames6.default)(prefixClass('checkbox-field'), _defineProperty({}, rootClassName, !!rootClassName), _defineProperty({}, className, !!className)) },
          radioArrayComponent
        );
      }

      return finalJSX;
    }
  }]);

  return RadioField;
}(_react2.default.Component);

RadioField.defaultProps = {
  data: [],
  onChange: function onChange() {},

  groupListFlag: true,
  groupListArgument: {
    lineIndent: 0,
    itemIndent: 15
  },
  label: ''
};

// http://facebook.github.io/react/docs/reusable-components.html
RadioField.propTypes = {
  className: _react2.default.PropTypes.string,
  data: _react2.default.PropTypes.array,
  onChange: _react2.default.PropTypes.func,
  groupListFlag: _react2.default.PropTypes.bool,
  groupListArgument: _react2.default.PropTypes.object
};

RadioField.displayName = 'RadioField';

module.exports = RadioField;
