// @ali/tingle-note/dist/Note.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Note Component for tingle
 * @author Kuncheng Zhao, zhouqan.yezq
 *
 * Copyright 2014-2016, Tingle Team.
 * All rights reserved.
 */
var classnames = require('classnames');
var Icon = require('@ali/tingle-icon');
var Context = require('@ali/tingle-context');

var prefixClass = Context.prefixClass;

var Note = function (_React$Component) {
    _inherits(Note, _React$Component);

    function Note(props) {
        _classCallCheck(this, Note);

        var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(Note).call(this, props));

        _this.state = {
            closed: false
        };
        return _this;
    }

    _createClass(Note, [{
        key: 'handleClose',
        value: function handleClose(e) {
            this.props.closable && this.doClose();
            this.props.onClose.call(this, e);
        }
    }, {
        key: 'doClose',
        value: function doClose() {
            this.setState({
                closed: true
            });
        }

        /**
         * 当closable为true的时候，展现关闭模式
         * 当closable为false的时候，展现查看详情模式
         */

    }, {
        key: 'render',
        value: function render() {
            var t = this;
            var type = 'note-' + t.props.type;
            var html = React.createElement(
                'div',
                { ref: 'root', className: classnames(prefixClass('FBH FBAC'), _defineProperty({}, t.props.className, !!t.props.className), prefixClass('note')) },
                React.createElement(Icon, { name: type, width: 22, height: 22 }),
                React.createElement(
                    'div',
                    { className: prefixClass('note-message FB1 omit') },
                    this.props.message
                ),
                React.createElement(
                    'div',
                    { className: prefixClass('FBH note-close'), onClick: t.handleClose.bind(t) },
                    t.props.closeText,
                    React.createElement(Icon, { name: t.props.closable ? 'cross' : 'angle-right',
                        width: 20, height: 20 })
                )
            );

            return this.state.closed ? null : html;
        }
    }]);

    return Note;
}(React.Component);

Note.defaultProps = {
    type: 'message',
    closable: false,
    closeText: '',
    message: '',
    onClose: function onClose() {}
};

// http://facebook.github.io/react/docs/reusable-components.html

Note.propTypes = {
    className: React.PropTypes.string,
    type: React.PropTypes.oneOf(['message', 'warning']),
    message: React.PropTypes.string,
    onClose: React.PropTypes.func,
    closeText: React.PropTypes.string
};

Note.displayName = 'Note';

module.exports = Note;
