// @ali/tingle/src/tingle.js
/**
 * Tingle All in One!
 * @author caoke.ck
 *
 * Copyright 2014-2015, Tingle Team, Alinw.
 * All rights reserved.
 */

var Context = require('tingle-context');

// 按照字母顺序排序
var Tingle = Context.mixin(Context, {
    Button:        require('tingle-button'),
    CityField:     require('tingle-city-field'),
    CityPicker:    require('tingle-city-picker'),
    Dialog:        require('tingle-dialog'),
    Field:         require('tingle-field'),
    GroupList:     require('tingle-group-list'),
    Icon:          require('tingle-icon'),
    Layer:         require('tingle-layer'),
    LoadMore:      require('tingle-load-more'),
    Mask:          require('tingle-mask'),
    NumberField:   require('tingle-number-field'),
    OnOff:         require('tingle-on-off'),
    OnOffField:    require('tingle-on-off-field'),
    PhotoField:    require('tingle-photo-field'),
    Picker:        require('tingle-picker'),
    Scroller:      require('tingle-scroller'),
    SearchBar:     require('tingle-search-bar'),
    SelectField:   require('tingle-select-field'),
    Slide:         require('tingle-slide'),
    Slot:          require('tingle-slot'),
    Tab:           require('tingle-tab'),
    TextareaField: require('tingle-textarea-field'),
    TextField:     require('tingle-text-field'),
    Tip:           require('tingle-tip')
});

module.exports = Tingle;
