// @ali/tingle-button/src/Button.js
/**
* Button Component for tingle
* @author cm
*
* Copyright 2014-2016, Tingle Team, Alinw.
* All rights reserved.
*/
const React = require('react');
const classnames = require('classnames');
const Context = require('@ali/tingle-context');

let prefixClass = function(name) {
    return Context.prefixClass ? Context.prefixClass(name) : 't-' + name
}

class Button extends React.Component {

    constructor(props) {
        super(props);
    }

    handleClick(evt) {
        if (!this.props.disabled) {
            this.props.onClick(evt);
        }
    }

    render() {
        let size = this.props.size

        let type = this.props.type
        let primary = type === 'primary'
        let secondary = type === 'secondary'
        let danger = type === 'danger'
        let text = type === 'text'
        let minor = type === 'minor'

        let disabled = !!this.props.disabled
        let classSet = {
            [`${prefixClass('FBH FBAC FBJC')}`]: true,
            [this.props.className]: !!this.props.className,
            [`${prefixClass('button')}`]: true,
            'disabled': disabled,
            [`${prefixClass('button-small')}`]: size === 'small',
            [`${prefixClass('button-large')}`]: size === 'large',
            [`${prefixClass('button-primary')}`]: primary && !disabled,
            [`${prefixClass('button-secondary')}`]: secondary && !disabled,
            [`${prefixClass('button-minor')}`]: minor && !disabled,
            [`${prefixClass('button-danger')}`]: danger && !disabled,
            [`${prefixClass('button-text')}`]: text
        }

        let style = this.props.style;

        return (
            <div className={classnames(classSet)}
                disabled={disabled}
                style={style}
                onClick={this.handleClick.bind(this)}>
                {this.props.children}
            </div>
        )
    }
}

Button.defaultProps = {
    size: 'medium',
    type: 'primary',
    disabled: false,
    style: {},
    onClick() {}
}

// http://facebook.github.io/react/docs/reusable-components.html
Button.propTypes = {
    size: React.PropTypes.oneOf(['medium', 'large', 'small']),
    type: React.PropTypes.oneOf(['primary', 'secondary', 'minor', 'danger', 'text']),
    disabled: React.PropTypes.bool,
    style: React.PropTypes.object,
    onClick: React.PropTypes.func
};

Button.displayName = 'Button';

module.exports = Button;
