// @ali/tingle-button/src/prototype.js

const React = require('react');
const {Bundle} = require('engine');
const {BoolSetter, TextSetter, ChoiceSetter, JsonSetter, NumberSetter} = require('engine-utils');

module.exports = Bundle.createPrototype({
    title: " 按钮",
    category: "*",
    icon: require("./logo-ide.svg"),
    componentName: "Button",
    canHovering: true,
    canSelecting: true,
    canDraging: true,
    isInline: false,
    isContainer: false,
    canDropto: true,
    canDroping: false,
    configure: [{
            name: "size",
            title: "size",
            defaultValue: 'medium',
            required: false,
            fieldStyle: "block",
            fieldCollapsed: false,
            setter: <ChoiceSetter options={[{value:'medium'},{value:'large'},{value:'small'}]} />
        }, {
            name: "type",
            title: "type",
            defaultValue: 'primary',
            required: false,
            fieldStyle: "block",
            fieldCollapsed: false,
            setter: <ChoiceSetter options={[{value:'primary'},{value:'secondary'},{value:'minor'},{value:'danger'},{value:'text'}]} />
        }, {
            name: "disabled",
            title: "disabled",
            defaultValue: false,
            fieldStyle: "block",
            fieldCollapsed: false,
            setter: <BoolSetter />
        }, {
            name: "style",
            title: "style",
            defaultValue: {},
            required: false,
            fieldStyle: "accordion",
            fieldCollapsed: false,
            setter: <JsonSetter />
        }, {
            name: "children",
            title: "text",
            defaultValue: "一级按钮",
            required: false,
            fieldStyle: "block",
            fieldCollapsed: false,
            setter: <TextSetter />
        }]
});
