// @ali/tingle-photo-field/dist/PhotoFieldItem.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var React = require('react');
var Context = require('@ali/tingle-context');
var prefixClass = Context.prefixClass;

var PhotoFieldItem = function (_React$Component) {
    _inherits(PhotoFieldItem, _React$Component);

    function PhotoFieldItem() {
        _classCallCheck(this, PhotoFieldItem);

        return _possibleConstructorReturn(this, (PhotoFieldItem.__proto__ || Object.getPrototypeOf(PhotoFieldItem)).apply(this, arguments));
    }

    _createClass(PhotoFieldItem, [{
        key: 'handlePreview',
        value: function handlePreview() {
            this.props.onPreviewImage(this.props.index);
        }
    }, {
        key: 'handleDelete',
        value: function handleDelete() {
            this.props.onDeleteImage(this.props.index);
        }
    }, {
        key: 'render',
        value: function render() {
            var t = this;
            return React.createElement(
                'div',
                { className: prefixClass('PR FL photo-item'), style: { width: t.props.width } },
                React.createElement('img', { src: '//gw.alicdn.com/tps/TB18GJsIpXXXXatXFXXXXXXXXXX.png',
                    preview: t.props.pic,
                    style: { backgroundImage: 'url(' + t.props.url + ')' }, alt: t.props.name,
                    onClick: t.handlePreview.bind(t)
                }),
                !t.props.readOnly && React.createElement('div', { className: prefixClass('PA photo-delete'), onClick: t.handleDelete.bind(t) })
            );
        }
    }]);

    return PhotoFieldItem;
}(React.Component);

PhotoFieldItem.defaultProps = {
    readOnly: false
};

PhotoFieldItem.propTypes = {
    index: React.PropTypes.number,
    readOnly: React.PropTypes.bool,
    onPreviewImage: React.PropTypes.func,
    onDeleteImage: React.PropTypes.func
};

module.exports = PhotoFieldItem;
