// @ali/tingle-photo-field/dist/PhotoField.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * PhotoField Component for tingle
 * @author
 * alex.mm
 * Copyright 2014-2016, Tingle Team.
 * All rights reserved.
 */

/* eslint-disable */
var React = require('react');
var classnames = require('classnames');

var Context = require('@ali/tingle-context');
var PhotoFieldItem = require('./PhotoFieldItem');
var Icon = require('@ali/tingle-icon');

var noop = function noop() {};

var autoFixUrl = function autoFixUrl(url) {
    var _url = url;
    if (_url) {
        // auto fix cdn url
        if (_url.indexOf('//') === 0) {
            _url = '' + location.protocol + _url;
        }
    }
    return _url;
};

var prefixClass = Context.prefixClass;

var Ali = window.Ali || {};

var PhotoField = function (_React$Component) {
    _inherits(PhotoField, _React$Component);

    function PhotoField(props) {
        _classCallCheck(this, PhotoField);

        var _this = _possibleConstructorReturn(this, (PhotoField.__proto__ || Object.getPrototypeOf(PhotoField)).call(this, props));

        _this.state = {
            width: 100 / props.columns + '%'
        };
        return _this;
    }

    _createClass(PhotoField, [{
        key: 'onPickHandler',
        value: function onPickHandler() {
            var _this2 = this;

            if (this.props.readOnly) {
                return;
            }
            var option = {
                multiple: !!this.props.multiple,
                max: this.props.max
            };
            if (Ali.isDingDing) {
                if (!this.props.corpId) {
                    Ali.alert({
                        message: this.props.locale === 'zh_CN' ? 'corpId 不能为空' : 'corpId is required',
                        okButton: this.props.locale === 'zh_CN' ? '确定' : 'OK'
                    });
                    return;
                }
                option.corpId = this.props.corpId;
                Ali.photo(option, function (result) {
                    if (result && !result.errorCode) {
                        _this2.props.onChange({
                            value: result.photos
                        });
                    } else {
                        Ali.alert({
                            message: result.errorMessage,
                            okButton: _this2.props.locale === 'zh_CN' ? '确定' : 'OK'
                        });
                    }
                });
            }
        }
    }, {
        key: 'renderPhotoList',
        value: function renderPhotoList() {
            var t = this;
            var photoItem = t.props.photoList.map(function (item, index) {
                var props = {
                    index: index,
                    url: item.url,
                    name: item.name,
                    width: t.state.width,
                    columns: t.props.columns,
                    readOnly: t.props.readOnly,
                    onDeleteImage: t.handleDeleteImage.bind(t),
                    onPreviewImage: t.handlePreview.bind(t)
                };
                return React.createElement(PhotoFieldItem, _extends({}, props, { key: index }));
            });
            return React.createElement(
                'div',
                { className: classnames(prefixClass('photo-list'), _defineProperty({}, prefixClass('hide'), !photoItem.length)) },
                photoItem
            );
        }
    }, {
        key: 'handleDeleteImage',
        value: function handleDeleteImage(index) {
            this.props.onDelete(index);
        }

        // 点击预览

    }, {
        key: 'handlePreview',
        value: function handlePreview(index) {
            var t = this;
            var urls = t.props.photoList.map(function (item) {
                var u = autoFixUrl(item.url);
                return { u: u };
            });

            Ali.imageViewer({
                images: urls,
                init: index
            }, function (result) {
                console.log(result);
            });
        }
    }, {
        key: 'render',
        value: function render() {
            var _classnames2;

            var t = this;
            var requiredTag = React.createElement(Icon, { name: 'field-required', className: prefixClass('field-layout-label-required'), width: 6, height: 6, fill: 'red' });
            return React.createElement(
                'div',
                { className: classnames(prefixClass('photo-field'), (_classnames2 = {}, _defineProperty(_classnames2, t.props.className, !!t.props.className), _defineProperty(_classnames2, prefixClass('hide'), t.props.hide), _defineProperty(_classnames2, 'readOnly', t.props.readOnly), _classnames2)) },
                t.props.label && t.props.layout === 'v' && React.createElement(
                    'div',
                    { className: prefixClass('field-layout-v-label') },
                    t.props.label,
                    this.props.required && requiredTag
                ),
                React.createElement(
                    'div',
                    { onClick: t.onPickHandler.bind(t), className: classnames(prefixClass('field-box FBH'), _defineProperty({}, prefixClass('TE'), !!t.props.tappable)) },
                    t.props.label && t.props.layout === 'h' && React.createElement(
                        'div',
                        { className: prefixClass('field-layout-h-label') },
                        t.props.label,
                        this.props.required && requiredTag
                    ),
                    React.createElement(
                        'div',
                        { className: prefixClass('FB1 PR FBH FBAC') },
                        React.createElement(
                            'div',
                            { className: classnames(prefixClass('omit select-field-placeholder'), _defineProperty({}, prefixClass('hide'), this.props.photoList.length || t.props.readOnly)) },
                            this.props.placeholder
                        )
                    ),
                    !t.props.readOnly && Ali.isDingDing && React.createElement(
                        'div',
                        { className: prefixClass('icon FBH FBAC photo-icon') },
                        React.createElement(Icon, { className: prefixClass('icon-svg'), name: 'photo', fill: t.props.fill, width: 26, height: 26 })
                    )
                ),
                this.renderPhotoList()
            );
        }
    }]);

    return PhotoField;
}(React.Component);

PhotoField.defaultProps = {
    columns: 3,
    multiple: true, // 是否同时上传多张照片
    layout: 'h',
    label: '',
    max: 9, // 钉钉组件一次性最多选择上传3张照片
    maxUpload: 12, // 总共上传图片总数
    readOnly: false,
    onChange: noop,
    photoList: [],
    fill: '#ccc',
    locale: 'zh_CN'
};

// http://facebook.github.io/react/docs/reusable-components.html
PhotoField.propTypes = {
    icon: React.PropTypes.object,
    columns: React.PropTypes.number,
    multiple: React.PropTypes.bool,
    corpId: React.PropTypes.string,
    placeholder: React.PropTypes.string,
    locale: React.PropTypes.string,
    label: React.PropTypes.string,
    max: React.PropTypes.number,
    maxUpload: React.PropTypes.number,
    photoList: React.PropTypes.array,
    readOnly: React.PropTypes.bool,
    onChange: React.PropTypes.func,
    onDelete: React.PropTypes.func
};

PhotoField.displayName = 'PhotoField';

module.exports = PhotoField;
