// @ali/tingle-avatar/src/Avatar.js
/**
 * Avatar Component for tingle
 * @author
 *
 * Copyright 2014-2016, Tingle Team.
 * All rights reserved.
 */
const classnames = require('classnames');
const { unitize } = require('@ali/tingle-style');
const Context = require('@ali/tingle-context');
const avatarColors = ["#FF8170", "#F5B153", "#56c8e8", "#79CC70", "#52bafa", "#9e9ee6"];
const defaultSrc = 'https://img.alicdn.com/tps/TB1.IgIKpXXXXbgXpXXXXXXXXXX-116-116.png';
let global = Context.getGlobal('avatar') || {};
class Avatar extends React.Component {

    constructor(props) {
        super(props);
        let size = unitize(props.size);
        this.style = {
            width: size,
            height: size,
            lineHeight: size,
            fontSize: unitize(props.fontSize)
        };
    }

    shouldComponentUpdate(nextProps, nextState) {
        for (let key in this.props) {
            if (this.props[key] !== nextProps[key]) {
                return true;
            }
        }
        return false;
    }

    render() {

        let t = this;
        if (!t.props.name || t.props.src) {
            return (
                <img ref='root' className={classnames(Context.prefixClass('avatar'), {
                    [t.props.className]: !!t.props.className
                })} src={t.props.src || t.props.defaultSrc} style={t.style}/>
            );
        }
        if (t.props.defaultColor) {
            t.style.backgroundColor = t.props.defaultColor;
        } else {
            let hashCode = t.props.hashCode(t.props.name, t.props.isLong);
            t.style.backgroundColor = t.props.colors[Math.abs(hashCode) % t.props.colors.length];
        }
        return (
            <div ref='root' className={classnames(Context.prefixClass('avatar'), {
                [t.props.className]: !!t.props.className
            })} style={this.style}>
                {Avatar.formatName(t.props.name)}
            </div>
        );
    }
}

Avatar.hashCode = (strKey, isLong) => {
    let intValue = (num) => {
        let MAX_VALUE = 0x7fffffff;
        let MIN_VALUE = -0x80000000;
        if (num > MAX_VALUE || num < MIN_VALUE) {
            return num &= 0xFFFFFFFF;
        }
        return num;
    };

    let hash = 0;
    if (typeof strKey === 'string') {
        for (let i = 0, l = strKey.length; i < l; i++) {
            hash = hash * 31 + strKey.charCodeAt(i);
            if (!isLong) {
                hash = intValue(hash);
            }
        }
    }
    return hash;
};

Avatar.formatName = (name) => {
    let firstChar = name.slice(0, 1);
    if (/^[A-Za-z]+$/.test(firstChar)) {
        return firstChar.toUpperCase();
    }
    if (name.length === 3) {
        return name.slice(1, 3);
    }
    if (name.length > 3) {
        return name.slice(0, 2);
    }
    return name;
};

Avatar.defaultProps = {
    className: '',
    name: '',
    src: '',
    defaultColor: '',
    colors: global.colors || avatarColors,
    defaultSrc: global.defaultSrc || defaultSrc,
    hashCode: global.hashCode || Avatar.hashCode,
    isLong: false,
    size: 40,
    fontSize: 12
};

// http://facebook.github.io/react/docs/reusable-components.html
Avatar.propTypes = {
    className: React.PropTypes.string,
    size: React.PropTypes.oneOfType([
        React.PropTypes.string,
        React.PropTypes.number
    ]),
    fontSize: React.PropTypes.oneOfType([
        React.PropTypes.string,
        React.PropTypes.number
    ]),
    name: React.PropTypes.string,
    defaultColor: React.PropTypes.string,
    defaultSrc: React.PropTypes.string,
    src: React.PropTypes.string,
    colors: React.PropTypes.array,
    isLong: React.PropTypes.bool,
    hashCode: React.PropTypes.func
};

Avatar.displayName = 'Avatar';

module.exports = Avatar;
