// @ali/tingle-photo-field/src/PhotoFieldItem.js
const React = require('react');
const Context = require('@ali/tingle-context');
const prefixClass = Context.prefixClass;

class PhotoFieldItem extends React.Component {

    handlePreview() {
        this.props.onPreviewImage(this.props.index);
    }

    handleDelete() {
        this.props.onDeleteImage(this.props.index);
    }

    render() {
        const t = this;
        return (
            <div className={prefixClass('PR FL photo-item')} style={{ width: t.props.width }}>
                <img src="//gw.alicdn.com/tps/TB18GJsIpXXXXatXFXXXXXXXXXX.png"
                     preview={t.props.pic}
                     style={{ backgroundImage: `url(${t.props.url})` }} alt={t.props.name}
                     onClick={t.handlePreview.bind(t)}
                />
                {
                    !t.props.readOnly &&
                    <div className={prefixClass('PA photo-delete')} onClick={t.handleDelete.bind(t)}/>
                }

            </div>
        );
    }
}

PhotoFieldItem.defaultProps = {
    readOnly: false,
};

PhotoFieldItem.propTypes = {
    index: React.PropTypes.number,
    readOnly: React.PropTypes.bool,
    onPreviewImage: React.PropTypes.func,
    onDeleteImage: React.PropTypes.func,
};

module.exports = PhotoFieldItem;
