// @ali/tingle-slot/dist/Slot.js
'use strict';

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Slot Component for tingle
 * @author caoke.ck
 *
 * Copyright 2014-2016, Tingle Team.
 * All rights reserved.
 */

var React = require('react');
var classnames = require('classnames');
var Context = require('@ali/tingle-context');
var Layer = require('@ali/tingle-layer');
var Scroller = require('@ali/tingle-scroller');

// 滑动效果的动画函数
var LINEAR_EASE = {
    style: 'linear',
    fn: function fn(k) {
        return k;
    }
};

var isArray = function isArray(arr) {
    return Object.prototype.toString.call(arr) === '[object Array]';
};
var equals = function equals(obj1, obj2) {
    return JSON.stringify(obj1) === JSON.stringify(obj2);
};

var Slot = function (_React$Component) {
    _inherits(Slot, _React$Component);

    function Slot(props) {
        _classCallCheck(this, Slot);

        var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(Slot).call(this, props));

        var t = _this;

        // 初始状态
        t.state = {
            visible: false,
            title: props.title || '',
            data: props.data || [],
            selectedIndex: t.findSelectedIndex(props)
        };
        return _this;
    }

    _createClass(Slot, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
            var t = this;

            // 获取所有 scroller 的容器
            var slotBody = t.refs.root.querySelector('.' + Context.prefixClass('slot-body'));

            // 获取选项高度
            t.findItemHeight(slotBody);

            // tap 事件触发选中状态变更
            slotBody.addEventListener('iscroll:tap', function (e) {
                var className = e.target.className;
                var reg = new RegExp(Context.prefixClass('slot-item(\\d+)_(\\d+)'));
                var match = reg.exec(className);
                if (match && className.indexOf(Context.prefixClass('slot-item-active')) === -1) {
                    var column = parseInt(match[1]);
                    var index = parseInt(match[2]);
                    t.props.onChange(t.getData(column, index), column, index);
                }
            }, false);

            // 初始化滚动的标记
            t._willRefresh = true;
        }
    }, {
        key: 'componentDidUpdate',
        value: function componentDidUpdate() {
            var t = this;

            // 获取选项高度
            t.findItemHeight();

            // 可见的时候滚动到选中的选项
            if (t.state.visible && t._willRefresh) {

                t._willRefresh = false;
                t.scrollAll(200);
            }
        }

        // 减少渲染次数

    }, {
        key: 'componentWillReceiveProps',
        value: function componentWillReceiveProps(nextProps) {
            var t = this;

            var data = nextProps.data;
            var selectedIndex = t.findSelectedIndex(nextProps);

            // 数据变化需要重新初始化 scroller
            var state = {};
            var willRefresh = false;
            if (!equals(t.state.data, data)) {
                state.data = data;
                if (t.props.scrollMod == 'keep') {
                    // 替换列后仍保留指定值的位置
                    // 记录旧值在新数据中的索引
                    t._selectedIndex = t.findSelectedIndex({
                        data: data,
                        value: t.state.selectedIndex.map(function (n, i) {
                            return t.state.data[i][n];
                        })
                    });
                    // 标记变更的列
                    t._columnChanged = t.state.data.map(function (n, i) {
                        return !equals(data[i], n);
                    });
                }
                willRefresh = true;
            }
            if (!equals(t.state.selectedIndex, selectedIndex)) {
                state.selectedIndex = selectedIndex;
                willRefresh = true;
            }
            if (willRefresh) {
                t._willRefresh = true;
                t.setState(state);
            }
        }
    }, {
        key: 'findItemHeight',
        value: function findItemHeight(slotBody) {
            var t = this;
            if (!t._itemHeight) {
                slotBody = slotBody || t.refs.root.querySelector('.' + Context.prefixClass('slot-body'));
                var li = slotBody.querySelector('li');
                t._itemHeight = li ? parseFloat(getComputedStyle(li, null).height) : 0;
            }
        }
    }, {
        key: 'scrollAll',
        value: function scrollAll(time) {
            var t = this;
            if (t.props.scrollMod == 'keep' && t._selectedIndex) {
                t._selectedIndex.forEach(function (index, column) {
                    var scroller = t.refs['scroller' + column].scroller;
                    if (t._columnChanged[column]) {
                        scroller.scrollTo(0, -index * t._itemHeight, 0, LINEAR_EASE);
                    }
                });
                delete t._selectedIndex;
                setTimeout(function () {
                    t.state.selectedIndex.forEach(function (index, column) {
                        var scroller = t.refs['scroller' + column].scroller;
                        scroller.scrollTo(0, -index * t._itemHeight, time, LINEAR_EASE);
                    });
                }, 5);
            } else {
                t.state.selectedIndex.forEach(function (index, column) {
                    var scroller = t.refs['scroller' + column].scroller;
                    scroller.scrollTo(0, -index * t._itemHeight, time, LINEAR_EASE);
                });
            }
        }
    }, {
        key: 'findSelectedIndex',
        value: function findSelectedIndex(props) {
            var data = props.data || [];
            var value = props.value || [];
            var selectedIndex = [];

            // 遍历数据模型
            data.forEach(function (columnData, column) {

                selectedIndex[column] = 0;

                // 遍历每一列
                for (var i = 0; i < columnData.length; i++) {

                    // 定位选中值
                    if (value[column] && columnData[i].value === value[column].value) {
                        selectedIndex[column] = i;
                        break;
                    }
                }
            });

            return selectedIndex;
        }
    }, {
        key: 'handleScrollStart',
        value: function handleScrollStart() {
            var t = this;
            t.setState({
                scrolling: true
            });
        }
    }, {
        key: 'handleScrollEnd',
        value: function handleScrollEnd(column) {
            var t = this;
            t.setState({
                scrolling: false
            });
            var scroller = t.refs['scroller' + column].scroller;
            var height = t._itemHeight;
            var remainder = Math.abs(scroller.y % height);
            var index = scroller.y / height;

            // 没有滚动到选项，需要继续滚动一段距离
            if (remainder) {

                var func = void 0;
                if (scroller.distY > 0) {
                    // 向下滚动
                    if (remainder < height * 0.7) {
                        func = 'ceil';
                    } else {
                        func = 'floor';
                    }
                } else {
                    // 向上滚动
                    if (remainder > height * 0.3) {
                        func = 'floor';
                    } else {
                        func = 'ceil';
                    }
                }

                index = Math[func](scroller.y / height);
            }

            // 在 onChange 中设置状态
            index = Math.abs(index);
            t.props.onChange(t.getData(column, index), column, index);
        }
    }, {
        key: 'show',
        value: function show() {
            var _this2 = this;

            this.setState({
                visible: true
            }, function () {
                _this2.scrollAll(0);
            });
        }
    }, {
        key: 'hide',
        value: function hide() {
            this.setState({
                visible: false
            });
        }
    }, {
        key: 'handleCancel',
        value: function handleCancel() {
            var t = this;
            if (t.state.scrolling) {
                return false;
            }

            try {
                t.props.onCancel();
            } finally {
                t.hide();
            }
        }
    }, {
        key: 'handleConfirm',
        value: function handleConfirm() {
            var t = this;
            if (t.state.scrolling) {
                return false;
            }

            try {
                t.props.onConfirm(t.getData());
            } finally {
                t.hide();
            }
        }

        // 获取值的时候指定变更的列，为什么要这么做，是因为有变更后我不直接改 state！

    }, {
        key: 'getData',
        value: function getData(sColumn, sIndex) {
            var t = this;
            var ret = [];
            var _t$state = t.state;
            var data = _t$state.data;
            var selectedIndex = _t$state.selectedIndex;


            selectedIndex.forEach(function (index, column) {
                ret[column] = data[column][column === sColumn ? sIndex : index];
            });

            return ret;
        }
    }, {
        key: 'render',
        value: function render() {
            var t = this;
            var _t$props = t.props;
            var className = _t$props.className;
            var maskCloseable = _t$props.maskCloseable;
            var cancelText = _t$props.cancelText;
            var confirmText = _t$props.confirmText;

            return React.createElement(
                Layer,
                { visible: t.state.visible, bottom: 0, onMaskClick: maskCloseable ? t.handleCancel.bind(t) : function () {
                        return maskCloseable;
                    } },
                React.createElement(
                    'div',
                    { ref: 'root', className: classnames(Context.prefixClass('slot'), _defineProperty({}, className, !!className)) },
                    React.createElement(
                        'div',
                        { className: Context.prefixClass('slot-header FBH FBAC') },
                        React.createElement(
                            'div',
                            { className: Context.prefixClass('slot-cancel'), onClick: t.handleCancel.bind(t) },
                            cancelText
                        ),
                        React.createElement(
                            'div',
                            { className: Context.prefixClass('FB1 FAC slot-title') },
                            t.state.title
                        ),
                        React.createElement(
                            'div',
                            { className: classnames(Context.prefixClass('slot-confirm'), {
                                    enable: !t.state.scrolling
                                }), onClick: t.handleConfirm.bind(t) },
                            confirmText
                        )
                    ),
                    t.props.columns && t.props.columns.length ? React.createElement(
                        'ul',
                        { className: Context.prefixClass('slot-columns FBH') },
                        t.props.columns.map(function (c, i) {
                            return React.createElement(
                                'li',
                                { key: 'column' + i, className: Context.prefixClass('FB1 FAC') },
                                c
                            );
                        })
                    ) : '',
                    React.createElement(
                        'div',
                        { className: Context.prefixClass('slot-body FBH FC9 PR') },
                        t.state.data.map(function (m, i) {
                            return React.createElement(
                                Scroller,
                                { ref: 'scroller' + i, key: 'scroller' + i, className: Context.prefixClass('FB1'), autoRefresh: t.state.visible, tap: 'iscroll:tap', onScrollStart: t.handleScrollStart.bind(t, i), onScrollEnd: t.handleScrollEnd.bind(t, i) },
                                React.createElement(
                                    'ul',
                                    null,
                                    React.createElement('li', null),
                                    React.createElement('li', null),
                                    m.map(function (n, j) {
                                        return React.createElement(
                                            'li',
                                            { key: 'item' + i + '_' + j, className: classnames(Context.prefixClass('slot-item' + i + '_' + j), _defineProperty({}, Context.prefixClass('slot-item-active'), j === t.state.selectedIndex[i])) },
                                            n.text
                                        );
                                    }),
                                    React.createElement('li', null),
                                    React.createElement('li', null)
                                )
                            );
                        })
                    )
                )
            );
        }
    }]);

    return Slot;
}(React.Component);

Slot.defaultProps = {
    title: '',
    value: [],
    maskCloseable: true,
    data: [],
    className: '',
    confirmText: '完成',
    cancelText: '取消',
    onConfirm: function onConfirm() {},
    onCancel: function onCancel() {},
    onChange: function onChange() {},

    scrollMod: 'reset',
    columns: []
};

// http://facebook.github.io/react/docs/reusable-components.html
Slot.propTypes = {
    className: React.PropTypes.string,
    title: React.PropTypes.string,
    data: React.PropTypes.array.isRequired,
    value: React.PropTypes.array,
    maskCloseable: React.PropTypes.bool,
    confirmText: React.PropTypes.string,
    cancelText: React.PropTypes.string,
    onConfirm: React.PropTypes.func,
    onCancel: React.PropTypes.func,
    onChange: React.PropTypes.func,
    scrollMod: React.PropTypes.string,
    columns: React.PropTypes.array
};

// 格式化单列数据
Slot.formatColumnValue = function (columnData, value) {

    // 兼容简单选中值
    var columnValue = void 0;
    if (typeof value !== 'undefined') {
        if (value.hasOwnProperty('value')) {
            columnValue = value.value;
        } else {
            columnValue = value;
        }
    }
    value = undefined;

    // 遍历每一项
    for (var i = 0; i < columnData.length; i++) {
        var cell = columnData[i];

        // 兼容非对象的数据
        if ((typeof cell === 'undefined' ? 'undefined' : _typeof(cell)) !== 'object') {
            cell = columnData[i] = {
                text: cell,
                value: cell
            };
        }

        // 补全缺失数据
        if (!cell.hasOwnProperty('text')) {
            cell.text = cell.value;
        }
        if (!cell.hasOwnProperty('value')) {
            cell.value = cell.text;
        }

        // 定位选中值
        if (cell.value === columnValue) {
            value = cell;
        }
    }

    // 默认选中第一项
    if ((typeof value === 'undefined' ? 'undefined' : _typeof(value)) !== 'object') {
        value = columnData[0];
    }

    return {
        columnData: columnData,
        columnValue: value
    };
};

// 格式化多列数据
Slot.formatDataValue = function () {
    var data = arguments.length <= 0 || arguments[0] === undefined ? [] : arguments[0];
    var value = arguments.length <= 1 || arguments[1] === undefined ? [] : arguments[1];


    // 兼容单列数据的缩略写法
    if (!isArray(data[0])) {
        data = [data];
    }
    if (!isArray(value)) {
        value = [value];
    }

    // 遍历数据模型
    data.forEach(function (columnData, column) {

        // 格式化列数据
        var ret = Slot.formatColumnValue(columnData, value[column]);
        value[column] = ret.columnValue;
    });

    return {
        data: data,
        value: value
    };
};

Slot.displayName = 'Slot';

module.exports = Slot;
